## Import Libraries


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


## Load the dataset

In [2]:
# Load the dataset
raw_data = "spotify_subscription.csv"
df = pd.read_csv(raw_data)

In [3]:
df # will show whole data

user_id  age        country   device subscription_plan  \
0           1   31          INDIA  Android            family   
1           2   25        Germany      iOS              Free   
2           3   44  United States      iOS           premium   
3           4   52          INDIA  Android            family   
4           5   60  United States  Android            family   
...       ...  ...            ...      ...               ...   
9995     9996   52         france  Android           Student   
9996     9997   34         france      iOS           premium   
9997     9998   20         canada  Android           premium   
9998     9999   34             UK      Ios           PREMIUM   
9999    10000   37         france  Android           PREMIUM   

      subscription_length_months auto_renewal  monthly_active_days  \
0                             26            0                    1   
1                             27            1                   10   
2                             31            0                    0   
3                             49            0                   14   
4                             13            0                   18   
...                          ...          ...                  ...   
9995                          17            0                    9   
9996                          16            1                    6   
9997                          48            0                    7   
9998                          53            1                   13   
9999                          30            1                    5   

      avg_session_duration_min  num_songs_streamed payment_method  \
0                         92.0               128.0    Credit Card   
1                        110.6               785.0            NaN   
2                         53.4               881.0         crypto   
3                         27.9               616.0     Google Pay   
4                         24.4               826.0         crypto   
...                        ...                 ...            ...   
9995                      19.1                64.0     Google Pay   
9996                      59.7               659.0            NaN   
9997                     112.6               848.0    Credit Card   
9998                       6.0               693.0         paypal   
9999                     105.3               520.0     Google Pay   

      num_failed_payments discount_used    churn  
0                     0.0           Yes       No  
1                     2.0             0       No  
2                     NaN             1       No  
3                     1.0           Yes      Yes  
4                    -1.0            No      Yes  
...                   ...           ...      ...  
9995                 -1.0           Yes       No  
9996                 -1.0           NaN       No  
9997                  1.0             0       No  
9998                  0.0           NaN  churned  
9999                  NaN           Yes       No  

[10000 rows x 14 columns]

In [4]:
# to read first few lines of data
df.head()

user_id  age        country   device subscription_plan  \
0        1   31          INDIA  Android            family   
1        2   25        Germany      iOS              Free   
2        3   44  United States      iOS           premium   
3        4   52          INDIA  Android            family   
4        5   60  United States  Android            family   

   subscription_length_months auto_renewal  monthly_active_days  \
0                          26            0                    1   
1                          27            1                   10   
2                          31            0                    0   
3                          49            0                   14   
4                          13            0                   18   

   avg_session_duration_min  num_songs_streamed payment_method  \
0                      92.0               128.0    Credit Card   
1                     110.6               785.0            NaN   
2                      53.4               881.0         crypto   
3                      27.9               616.0     Google Pay   
4                      24.4               826.0         crypto   

   num_failed_payments discount_used churn  
0                  0.0           Yes    No  
1                  2.0             0    No  
2                  NaN             1    No  
3                  1.0           Yes   Yes  
4                 -1.0            No   Yes

In [5]:
# tells the number of rows and colums of the dataset
df.shape

(10000, 14)

In [6]:
df.describe()

user_id           age  subscription_length_months  \
count  10000.00000  10000.000000                10000.000000   
mean    5000.50000     40.799100                   29.787900   
std     2886.89568     13.619269                   16.981977   
min        1.00000     18.000000                    1.000000   
25%     2500.75000     29.000000                   15.000000   
50%     5000.50000     41.000000                   30.000000   
75%     7500.25000     53.000000                   45.000000   
max    10000.00000    150.000000                   59.000000   

       monthly_active_days  avg_session_duration_min  num_songs_streamed  \
count         10000.000000              10000.000000         9998.000000   
mean             14.439200                 62.946190          506.731646   
std               8.633965                 33.181852          399.333686   
min               0.000000                  5.000000            0.000000   
25%               7.000000                 33.400000          251.000000   
50%              14.000000                 63.550000          501.000000   
75%              22.000000                 91.400000          755.000000   
max              29.000000                120.000000        20000.000000   

       num_failed_payments  
count          7993.000000  
mean              0.510697  
std               1.118668  
min              -1.000000  
25%               0.000000  
50%               1.000000  
75%               2.000000  
max               2.000000

## Data Cleaning

In [7]:
# to check the datatyes in the dataset
df.dtypes

user_id                         int64
age                             int64
country                        object
device                         object
subscription_plan              object
subscription_length_months      int64
auto_renewal                   object
monthly_active_days             int64
avg_session_duration_min      float64
num_songs_streamed            float64
payment_method                 object
num_failed_payments           float64
discount_used                  object
churn                          object
dtype: object

In [8]:
# tells about information of each column of data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   user_id                     10000 non-null  int64  
 1   age                         10000 non-null  int64  
 2   country                     10000 non-null  object 
 3   device                      9887 non-null   object 
 4   subscription_plan           8317 non-null   object 
 5   subscription_length_months  10000 non-null  int64  
 6   auto_renewal                9816 non-null   object 
 7   monthly_active_days         10000 non-null  int64  
 8   avg_session_duration_min    10000 non-null  float64
 9   num_songs_streamed          9998 non-null   float64
 10  payment_method              8373 non-null   object 
 11  num_failed_payments         7993 non-null   float64
 12  discount_used               8040 non-null   object 
 13  churn                       9479

In [9]:
# Convert column names to keep first letter captal while other letters lowercase and keep the underscores
df.columns = df.columns.str.replace("_", " ").str.title().str.replace(" ", "_")

In [10]:
df.head()

User_Id  Age        Country   Device Subscription_Plan  \
0        1   31          INDIA  Android            family   
1        2   25        Germany      iOS              Free   
2        3   44  United States      iOS           premium   
3        4   52          INDIA  Android            family   
4        5   60  United States  Android            family   

   Subscription_Length_Months Auto_Renewal  Monthly_Active_Days  \
0                          26            0                    1   
1                          27            1                   10   
2                          31            0                    0   
3                          49            0                   14   
4                          13            0                   18   

   Avg_Session_Duration_Min  Num_Songs_Streamed Payment_Method  \
0                      92.0               128.0    Credit Card   
1                     110.6               785.0            NaN   
2                      53.4               881.0         crypto   
3                      27.9               616.0     Google Pay   
4                      24.4               826.0         crypto   

   Num_Failed_Payments Discount_Used Churn  
0                  0.0           Yes    No  
1                  2.0             0    No  
2                  NaN             1    No  
3                  1.0           Yes   Yes  
4                 -1.0            No   Yes

In [11]:
# Convert Auto_Renewal value to binary
df['Auto_Renewal'] = df['Auto_Renewal'].replace({'Yes' : 1, 'No' : 0})

# Handle missing values in Auto_Renewal column with most common values 
df['Auto_Renewal'].fillna(df['Auto_Renewal'].mode()[0], inplace=True)
df['Auto_Renewal'] = df['Auto_Renewal'].astype(int)

In [12]:
df.dtypes

User_Id                         int64
Age                             int64
Country                        object
Device                         object
Subscription_Plan              object
Subscription_Length_Months      int64
Auto_Renewal                    int32
Monthly_Active_Days             int64
Avg_Session_Duration_Min      float64
Num_Songs_Streamed            float64
Payment_Method                 object
Num_Failed_Payments           float64
Discount_Used                  object
Churn                          object
dtype: object

In [13]:
df['Auto_Renewal']

0       0
1       1
2       0
3       0
4       0
       ..
9995    0
9996    1
9997    0
9998    1
9999    1
Name: Auto_Renewal, Length: 10000, dtype: int32

In [14]:
# Find missing values
df.isna()

User_Id    Age  Country  Device  Subscription_Plan  \
0       False  False    False   False              False   
1       False  False    False   False              False   
2       False  False    False   False              False   
3       False  False    False   False              False   
4       False  False    False   False              False   
...       ...    ...      ...     ...                ...   
9995    False  False    False   False              False   
9996    False  False    False   False              False   
9997    False  False    False   False              False   
9998    False  False    False   False              False   
9999    False  False    False   False              False   

      Subscription_Length_Months  Auto_Renewal  Monthly_Active_Days  \
0                          False         False                False   
1                          False         False                False   
2                          False         False                False   
3                          False         False                False   
4                          False         False                False   
...                          ...           ...                  ...   
9995                       False         False                False   
9996                       False         False                False   
9997                       False         False                False   
9998                       False         False                False   
9999                       False         False                False   

      Avg_Session_Duration_Min  Num_Songs_Streamed  Payment_Method  \
0                        False               False           False   
1                        False               False            True   
2                        False               False           False   
3                        False               False           False   
4                        False               False           False   
...                        ...                 ...             ...   
9995                     False               False           False   
9996                     False               False            True   
9997                     False               False           False   
9998                     False               False           False   
9999                     False               False           False   

      Num_Failed_Payments  Discount_Used  Churn  
0                   False          False  False  
1                   False          False  False  
2                    True          False  False  
3                   False          False  False  
4                   False          False  False  
...                   ...            ...    ...  
9995                False          False  False  
9996                False           True  False  
9997                False          False  False  
9998                False           True  False  
9999                 True          False  False  

[10000 rows x 14 columns]

In [15]:
# To see how many missing values in each column
df.isna().sum()

User_Id                          0
Age                              0
Country                          0
Device                         113
Subscription_Plan             1683
Subscription_Length_Months       0
Auto_Renewal                     0
Monthly_Active_Days              0
Avg_Session_Duration_Min         0
Num_Songs_Streamed               2
Payment_Method                1627
Num_Failed_Payments           2007
Discount_Used                 1960
Churn                          521
dtype: int64

In [16]:
# Display rows which has missing values
df[df.isna().any(axis=1)]

User_Id  Age        Country   Device Subscription_Plan  \
1           2   25        Germany      iOS              Free   
2           3   44  United States      iOS           premium   
6           7   44        Germany      NaN           Student   
8           9   50          INDIA      iOS           PREMIUM   
9          10   48          INDIA      iOS            family   
...       ...  ...            ...      ...               ...   
9993     9994   27          india  Android           premium   
9994     9995   49  United States  Desktop           premium   
9996     9997   34         france      iOS           premium   
9998     9999   34             UK      Ios           PREMIUM   
9999    10000   37         france  Android           PREMIUM   

      Subscription_Length_Months  Auto_Renewal  Monthly_Active_Days  \
1                             27             1                   10   
2                             31             0                    0   
6                             38             1                    8   
8                             14             1                    3   
9                             56             0                   17   
...                          ...           ...                  ...   
9993                          13             0                   14   
9994                          41             0                    9   
9996                          16             1                    6   
9998                          53             1                   13   
9999                          30             1                    5   

      Avg_Session_Duration_Min  Num_Songs_Streamed Payment_Method  \
1                        110.6               785.0            NaN   
2                         53.4               881.0         crypto   
6                        110.8               307.0            NaN   
8                         14.3               394.0            NaN   
9                        115.5               550.0            NaN   
...                        ...                 ...            ...   
9993                      82.6               233.0    Credit Card   
9994                      59.2               661.0         crypto   
9996                      59.7               659.0            NaN   
9998                       6.0               693.0         paypal   
9999                     105.3               520.0     Google Pay   

      Num_Failed_Payments Discount_Used    Churn  
1                     2.0             0       No  
2                     NaN             1       No  
6                    -1.0           Yes      Yes  
8                     0.0             1      Yes  
9                    -1.0             0      Yes  
...                   ...           ...      ...  
9993                  NaN            No       No  
9994                  1.0           NaN      NaN  
9996                 -1.0           NaN       No  
9998                  0.0           NaN  churned  
9999                  NaN           Yes       No  

[5813 rows x 14 columns]

In [17]:
# Display duplicate values
df.duplicated().sum()

0

In [18]:
# Standardize country names
country_names = {
        "Us": "United States",
        "Usa": "United States",
        "U.S.A.": "United States",
        "U.S.": "United States",
        "United States Of America": "United States"
}
df['Country'] = df['Country'].astype(str).str.strip().str.title()
df['Country'] = df['Country'].replace(country_names)

In [19]:
# Replacing missing values in Device column with 'Unknown'
df['Device'].fillna("Unknown", inplace = True)

In [20]:
df.head()

User_Id  Age        Country   Device Subscription_Plan  \
0        1   31          India  Android            family   
1        2   25        Germany      iOS              Free   
2        3   44  United States      iOS           premium   
3        4   52          India  Android            family   
4        5   60  United States  Android            family   

   Subscription_Length_Months  Auto_Renewal  Monthly_Active_Days  \
0                          26             0                    1   
1                          27             1                   10   
2                          31             0                    0   
3                          49             0                   14   
4                          13             0                   18   

   Avg_Session_Duration_Min  Num_Songs_Streamed Payment_Method  \
0                      92.0               128.0    Credit Card   
1                     110.6               785.0            NaN   
2                      53.4               881.0         crypto   
3                      27.9               616.0     Google Pay   
4                      24.4               826.0         crypto   

   Num_Failed_Payments Discount_Used Churn  
0                  0.0           Yes    No  
1                  2.0             0    No  
2                  NaN             1    No  
3                  1.0           Yes   Yes  
4                 -1.0            No   Yes

In [57]:
# Remove rows where more than 2 columns have NaN values
df = df[df.isnull().sum(axis=1) <=2]

In [23]:
# Reset Indexing
df.reset_index(drop=True, inplace=True)

In [24]:
# Handle missing values in Discount_Used column by filling with mode
# Step 1: Convert 'Yes'/'No' to binary values
df.loc[:, 'Discount_Used'] = df['Discount_Used'].replace({'Yes': 1, 'No': 0})

# Step 2: Fill missing values using mode (most common value)
discount_mode = df['Discount_Used'].mode()[0]
df.loc[:, 'Discount_Used'] = df['Discount_Used'].fillna(discount_mode).astype(int)


In [58]:
# Drop missing values in 'Churn'
df = df.dropna(subset=['Churn'])

In [59]:
# Standardize Churn column values
df['Churn'] = df['Churn'].astype(str).str.strip().str.capitalize()
df['Churn'] = df['Churn'].replace({'Y': 'Yes', 'N': 'No', 'Churned': 'Yes'})


In [28]:
# Reset index after row removal
df.reset_index(drop=True, inplace=True)

In [30]:
# Handle incorrect values in numerical columns
numerical_cols = ['Age', 'Subscription_Length_Months', 'Monthly_Active_Days', 'Avg_Session_Duration_Min']
for col in numerical_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce')  # Convert to numeric, setting invalid values to NaN

In [31]:
# Find nagative values
for col in numerical_cols:
    negative_rows = df[df[col] < 0]
    if not negative_rows.empty:
        print(f"Rows with negative values in {col}:")
        print(negative_rows[[col]])

In [32]:
for col in numerical_cols:
    df.loc[df[col] < 0, col] = np.nan

In [33]:
# Fix Num_Failed_Payments: Convert to numeric, handle negatives and NaNs using mode
df['Num_Failed_Payments'] = pd.to_numeric(df['Num_Failed_Payments'], errors='coerce')

# Replace negative values with NaN
df.loc[df['Num_Failed_Payments'] < 0, 'Num_Failed_Payments'] = np.nan

# Fill NaNs with mode
mode_failed_payments = df['Num_Failed_Payments'].mode()[0]
df['Num_Failed_Payments'].fillna(mode_failed_payments, inplace=True)

# Convert to integer
df['Num_Failed_Payments'] = df['Num_Failed_Payments'].astype(int)


In [35]:
# Fill missing 'Payment_Method' values with 'Unknown'
df['Payment_Method'].fillna('Unknown', inplace=True)


In [48]:
# Fill missing values with 'Unknown'
df['Subscription_Plan'].fillna('Unknown', inplace=True)

# Standardize case and trim spaces
df['Subscription_Plan'] = df['Subscription_Plan'].astype(str).str.strip().str.title()

# Map common misspellings and inconsistencies
plan_mapping = {
    'Premuim': 'Premium',
    'Primium': 'Premium',
    'Premium ': 'Premium',
    'Basik': 'Basic',
    'Famliy': 'Family',
    'Famly': 'Family',
}
df['Subscription_Plan'] = df['Subscription_Plan'].replace(plan_mapping)

In [50]:
# Drop rows for Num_Songs_Streamed which has missing values
df.dropna(subset=['Num_Songs_Streamed'], inplace=True)


In [51]:
# Convert Num_Songs_Streamed to numeric type
df['Num_Songs_Streamed'] = pd.to_numeric(df['Num_Songs_Streamed'], errors='coerce')

In [52]:
# Cap outliers above the 95th percentile
cap = df['Num_Songs_Streamed'].quantile(0.95)
df['Num_Songs_Streamed'] = np.where(df['Num_Songs_Streamed'] > cap, cap, df['Num_Songs_Streamed'])


In [53]:
df.isna().sum()

User_Id                       0
Age                           0
Country                       0
Device                        0
Subscription_Plan             0
Subscription_Length_Months    0
Auto_Renewal                  0
Monthly_Active_Days           0
Avg_Session_Duration_Min      0
Num_Songs_Streamed            0
Payment_Method                0
Num_Failed_Payments           0
Discount_Used                 0
Churn                         0
dtype: int64

In [56]:
# Save cleaned data
df.to_csv("cleaned_spotify_churn.csv", index=False)
print("\nCleaned dataset saved successfully as 'cleaned_spotify_churn.csv'")



Cleaned dataset saved successfully as 'cleaned_spotify_churn.csv'


In [61]:
# Save the dataset again as I had to change the Churn row. This will overwrite the previous version 
df.to_csv("cleaned_spotify_churn.csv", index=False)
print("Updated cleaned dataset saved successfully!")


Updated cleaned dataset saved successfully!


# Data Cleaning Summary

1. **What kind of cleaning steps did you perform**
      
- Standardized column names by capitalizing the first letter of each word.
- Removed duplicate rows to ensure data consistency.
- Cleaned and standardized categorical columns like Country, Subscription_Plan, Device, and Churn.
  - Fixed inconsistent formatting (e.g., 'premium', 'PREMIUM' → 'Premium')
  - Mapped spelling errors (e.g., 'Premuim', 'Primium' → 'Premium')
- Converted binary columns (Autorenewal, Discount_Used) to 0 and 1.
- Standardized the Churn column ('Y', 'N', 'churned' → 'Yes', 'No')
- Removed rows with more than 2 missing values.

2. **How did you deal with missing values, if there were any?**
- Categorical Columns:
  - Filled NaN in Country, Device, Subscription_Plan, and Payment_Method with 'Unknown'.
  - Filled NaN in Autorenewal and Discount_Used with the most frequent value (mode).
- Numerical Columns:
  - Replaced negative values in features like Age, Subscription_Length_Months, Avg_Session_Duration_Min, etc., with NaN, then filled using the median.
  - Filled missing values in Num_Failed_Payments with the mode.
  - Dropped 2 rows that had missing values in Num_Songs_Streamed.

3. **Were there outliers, and how did you handle them?**

Yes, outliers were detected and handled using capping:

- Age: Capped values > 100 to 65.

- Avg_Session_Duration_Min: Capped values > 300 to 120.

- Num_Songs_Streamed: Capped values above the 95th percentile to avoid extreme distortion.